### PR TITLE
Fix the uncatchable-abort family: record sizing, devfs stat, SRFI-18 timeouts, 255-arg calls

### LINKED ISSUE #1973: A record with 27 or more fields aborts the process on instantiation — u8 overflow in allocRecordInstance
**Category:** GC / records · **Severity:** crash · **Found by:** systematic audit v2, Phase 2.5 (#1890)

Creating an *instance* of a record type with **≥27 fields** panics with an integer overflow. Not catchable by `guard`; exit 134.

**This is plain R7RS `define-record-type`.** No SRFI 237, no procedural layer, nothing exotic.

## Reproduction

```scheme
(import (scheme base) (scheme write))
(define-record-type big (mk f0 f1 ... f26) big? (f0 g0) ... (f26 g26))
(big? (mk 0 1 ... 26))
```

```text
thread 32696792 panic: integer overflow
src/gc_alloc.zig:473 in allocRecordInstance
exit 134
```

Wrapping the constructor in `guard` does not help — still exit 134.

## Bisected cliff

| fields | result |
|---:|---|
| 26 | `#t` |
| **27** | **panic** |
| 28 | panic |

## Mechanism

`gc_alloc.zig:473`:

```zig
self.finishAlloc(&ri.header, @sizeOf(RecordInstance) + record_type.num_fields * @sizeOf(Value));
```

`num_fields` is **`u8`** (`types_record.zig:11`), so `num_fields * @sizeOf(Value)` is evaluated in `u8` arithmetic. 40 + 8×27 = 256, which overflows.

## The dangerous part

**The record *type* is created without complaint.** Only the first instantiation dies. So a library can define a wide record type, load fine, pass its own type-level tests, and abort the first time anyone constructs one.

Reachable three ways, all verified: own fields, an inherited chain (20 + 7), and plain R7RS as above.

## Inferred, and worth checking before fixing

Under `ReleaseFast`, unsigned overflow is illegal behaviour rather than a trap — `finishAlloc` would receive a **wrapped** byte count and silently corrupt GC accounting instead of aborting. Not verified (the repro was run under ReleaseSafe); it changes the severity from "crash" to "silent heap corruption" if true, so it should be established before choosing a fix.

## Suggested fix

Widen the arithmetic at the call site (`@as(usize, record_type.num_fields) * @sizeOf(Value)`), and decide separately whether `num_fields: u8` is the right type at all — the 255-field limit it implies is enforced elsewhere with a proper `argError`, so the field *count* cap is deliberate, but the size *computation* should not inherit the narrow type.

`tests/scheme/audit/primitives_srfi237-audit.scm` (Phase 2.5) has 3 assertions disabled against this.

**Verified:** fresh ReleaseSafe build, isolated `KAAPPI_HOME`; the 26/27/28 bisect, the exit code, the `guard` case, and the arithmetic all reproduced independently.

### LINKED ISSUE #1976: file-info aborts the process on any devfs path — signed st_dev @intCast into u64
**Category:** SRFI 170 filesystem · **Severity:** crash · **Found by:** systematic audit v2, Phase 2.12 (#1890)

`file-info` on anything under `/dev` aborts. Uncatchable by `guard`; exit 134.

```scheme
(guard (e (#t 'caught)) (file-info "/dev/null" #t))
```

```text
thread 33127508 panic: integer does not fit in destination type
exit 134
```

## Mechanism

`src/primitives_filesystem.zig:134` — `.dev = @intCast(stat_buf.dev)` narrows a **signed** `dev_t` into `u64`. macOS devfs reports `st_dev = 2998493978`, which is `−1296473318` as `i32`.

## Discriminating controls

| path | result | what it isolates |
|---|---|---|
| `/tmp` | `directory` | positive `st_dev` is fine — not "file-info is broken" |
| `/dev/null` | **abort** | character device |
| **`/dev/fd`** | **abort** | **a *directory* with `st_rdev = 0`** |

`/dev/fd` is the useful one: it is not a device node and has no `rdev`, yet it still aborts — so the fault is `st_dev`, not `rdev`, and not the file *type*.

Identical under `-Dgc-stress=true`, so not a GC interaction.

## Why this matters more than a device-node edge case

`/dev/null` is an ordinary path for ordinary code — a build script asking whether it exists, a test fixture, a config default. Any `file-info` on a devfs path takes down the process with no catchable error.

It also makes device nodes **untestable**: the audit could not assert anything about `file-info` on them, because reaching them aborts first.

## Platform scope

- **macOS: verified affected.**
- **OpenBSD: inferred affected** — `dev_t = i32`. Read from the type declaration, not run.
- Linux (`statx` + `makedev`, u64), Windows (`c_uint`), FreeBSD and NetBSD (`u64`) are safe by construction — also inferred from declarations.

## Suggested fix

`@bitCast` to the unsigned width before widening, or store `dev`/`rdev` as signed. Whichever is chosen, `rdev` at `:135` needs the same treatment — it is the same signed type and simply happens not to be negative on the paths that currently reach it.

**Verified:** fresh ReleaseSafe build at `03bda165`, isolated `KAAPPI_HOME`, cache cleared; the abort, its exit code, the `guard` case and all three controls reproduced independently.

### LINKED ISSUE #1983: Unguarded float→int conversions in SRFI-18 abort the process; timeoutToDeadlineNs takes (kaappi fibers) down too
**Category:** SRFI-18 / fibers · **Severity:** crash · **Found by:** systematic audit v2, Phase 5D (#1890)

```scheme
(guard (e (#t 'caught)) (seconds->time +inf.0))    ⟹ exit 134, uncatchable
```

Three sites, two panic kinds, each with a bisected cliff and a control:

| site | trigger | control |
|---|---|---|
| `secondsToTimeFn:1397` — `@intFromFloat(@floor(secs))` → i64 | `±inf.0`, `\|x\| ≥ 2^63` | `9.2e18` fine, **`9.3e18` aborts** |
| `threadSleepFn:604` → u64 | timeout ≳ 1.845e10 s | `1.8e10` catchable, **`1.9e10` aborts** |
| `timeoutToDeadlineNs:244` → u64, and `:234` (time-object branch, separate `"integer overflow"` panic) | same threshold | same |

Reproduced 5/5 across a 15-value matrix.

## The blast radius is wider than SRFI 18

`timeoutToDeadlineNs` is `pub`, and **`(kaappi fibers)` `channel-receive` aborts** on a `1e300` timeout. Control: `1.8e10` is catchable there too. So a fiber-based program with a large or computed timeout takes the process down with no catchable error.

## Same class as closed #604/#846 and open #1907

That family is unguarded numeric conversion at a boundary. None of *these* sites is filed, and the `timeoutToDeadlineNs` one is the first to reach beyond its own SRFI.

A timeout is exactly the argument most likely to be computed rather than written literally — `(* retries backoff)` overflowing is an ordinary bug that should raise, not abort.

**Verified:** fresh ReleaseSafe build, isolated `KAAPPI_HOME`; the `+inf.0` abort, both cliffs, and the `channel-receive` case reproduced with controls.

### LINKED ISSUE #2185: 255-argument calls and 256-parameter lambdas abort the process — u8 arithmetic in callClosure, tail dispatch, and compileLambda
**Category:** VM / compiler · **Severity:** crash · **Found by:** adjacent-site sweep while fixing #1973 (same unguarded-narrow-arithmetic family as #1973/#1976/#1983)

Two independent aborts, both uncatchable (exit 134), both from ordinary programs:

## 1. Calling a 255-parameter procedure aborts

```scheme
((lambda (p0 p1 ... p254) p254) 0 1 ... 254)   ; 255 params, 255 args
```

```text
thread NNN panic: integer overflow
src/vm_calls.zig callClosure
```

Bisected: 253 and 254 parameters fine, **255 aborts**. The bytecode ISA encodes `nargs` as a u8, so 255 is the largest argument count a call site can express — i.e. this is the *last legal* call, and it dies.

`vm_calls.zig:508`:

```zig
try vm.ensureRegisterCapacity(@as(usize, base) + @as(usize, @max(nargs + 1, func.locals_count)) + 1);
```

`nargs` is u8, so `nargs + 1` is u8 arithmetic: 255 + 1 overflows.

Three sibling sites compute `func.arity + 1` in u8 for the variadic-callee register window (rest-list slot after 255 fixed params): `vm_dispatch.zig:462` (tail_call), `:674` (tail apply, explicitly `arg_count: u8`), `:1079` (the tail-call reuse path). Reached by a variadic `(lambda (p0 ... p254 . r) ...)` — arity 255, `arity + 1` overflows.

## 2. Defining a 256-parameter lambda aborts at compile time

```scheme
(define f (lambda (p0 p1 ... p255) p0))   ; 256 params
```

```text
kaappi internal error — while: compiling ...
panic: integer overflow
```

`compiler_lambda.zig:78`: `arity += 1` with `arity: u8` — the 256th parameter overflows. `kaappi check` on such a file aborts the same way (static analysis dies rather than reporting).

Since no call site can ever supply 256 arguments (u8 `nargs`, and the apply path already rejects >255 loudly via `tooManyApplyArgs`), a >255-parameter lambda is uncallable by construction — the right behavior is a clean compile error, mirroring the 255-field cap `parseRecordSpec`/`allocRecordTypeExtended` enforce for records.

## Same class as #1973 / #1976 / #1983

Unguarded narrow arithmetic at a representation boundary. Found by sweeping for sibling sites of #1973's `num_fields * @sizeOf(Value)` u8 overflow.

**Verified:** fresh ReleaseSafe build on this branch; 253/254/255-param bisect and the 256-param compile abort reproduced; fix + regression tests in the same PR as the #1973/#1976/#1983 fixes.

### PR DESCRIPTION
**Pull request by baijum**
Fixes four issues sharing one root cause — unguarded narrow arithmetic or conversion at a representation boundary, each an exit-134 process abort reachable from an ordinary program and invisible to `guard`. One PR per the request; #1907 turned out to already be fixed and is verified here.

## #1973 — record with ≥27 fields aborts on instantiation

`allocRecordInstance` sized its allocation as `num_fields * @sizeOf(Value)` in **u8 arithmetic** (`num_fields` is a u8), overflowing at 27 fields (40 + 8·27 = 256). Widened to `usize`. On the issue's ReleaseFast question: confirmed the overflow would wrap there instead of trapping, silently corrupting GC accounting — the widening fixes both modes, and `gc_sweep.objectSize` (which sizes the same object from `fields.len`, a usize) now agrees with the alloc side.

Adjacent hole closed: `parseRecordSpec` admitted **256-field** specs whose later `@intCast` into the u8 aborted at *definition* time; capped at 255 to match the R6RS parser. The construction ceiling is pinned honestly: **254 fields** is the widest constructible record everywhere (both the syntactic desugar and SRFI-237's `record-constructor` pass rtd+fields through one call, and the call/apply ISA caps at 255 arguments), with 255-field construction a loud KP3007. The three audit assertions disabled against this bug are re-enabled.

## #1976 — `file-info` aborts on devfs paths

`doStat` `@intCast` a signed `dev_t` (i32 on macOS/OpenBSD) into u64; macOS `/dev/null` stats as `st_dev = -1296473318`. New `devToU64` reinterprets the bits as the unsigned same-width type before widening (a device id is an opaque identity, not a quantity), applied to `dev` and `rdev` on both stat branches per the issue's note that `rdev` is the same type and merely lucky. `file-info:device` on `/dev/null` now reports the issue's exact `2998493978`. The audit's disabled devfs assertions (including the discriminating `/dev/fd` directory case) are re-enabled.

## #1983 — SRFI-18 float→int conversions abort; `(kaappi fibers)` too

Two deliberate post-fix behaviors:

- **`seconds->time` raises** a catchable KP3007 outside [-2^63, 2^63), including ±inf.0 and NaN — it constructs an observable value, so clamping would silently build a wrong time. The bound is written `0x1p63` exactly; `@floatFromInt(maxInt(i64))` rounds *up* to 2^63 and would re-admit the first aborting value — the same off-by-one-ULP guard shape #1907 was made of.
- **Timeouts saturate** (`*|`, `+|`, `std.math.lossyCast`): +inf.0 and any beyond-range duration or far-future time object mean "never times out" — the SRFI-18 convention (Gambit's own reading of +inf.0), continuous with `#f`, and it keeps legal far-future time objects (i64 seconds reach ~year 2554+ wall time) usable as deadlines rather than raising on them. NaN stays "no wait", as before. `timeoutToDeadlineNs` is `pub` and shared with `(kaappi fibers)`, so channel-send/receive timeouts are fixed by the same change — the issue's `channel-receive` 1e300 case is a test now.

The audit file's disabled abort cases are re-enabled under this contract (note: its timeout blocks were marked `FAIL: #1984`, but that content matches #1983's sites — the real #1984 is the mutex/terminate state-machine issue and is untouched here). Its stale "TODAY: NaN does not abort (value undefined)" pin — the one suite failure this PR's changes initially caused — is updated to the new catchable rejection.

## #2185 — 255-argument calls and 256-parameter lambdas abort (new issue, found here)

Filed and fixed in this PR: sweeping for sibling sites of #1973's pattern found that `callClosure` computed `nargs + 1` in u8 — aborting **every 255-argument call, the ISA's own maximum** (253/254 worked, which is how it shipped); three tail-dispatch paths computed a variadic callee's `arity + 1` rest-slot count the same way; and both `compileLambda` copies (`compiler_lambda.zig` and the live IR path in `compiler_ir.zig`) overflowed their u8 arity counter on a **256-parameter lambda at compile time** — killing `kaappi check` too. Calls widened to usize; a 256th fixed parameter is now a clean KP2001 (`InvalidSyntax`, deliberately not `TooManyLocals`, which surfaces as a KP9001 "internal error — report this bug" and would blame kaappi for user syntax).

## #1907 — verified fixed, no code needed

The reader `#e`-at-2^63 panic was fixed by #2181 (merged): verified on this branch that `read`, `kaappi check`, `kaappi ast`, and `kaappi fmt` all handle `#e9223372036854775808.0`, and both boundary values (`…296.0` first-crashing, `…784.0` last-safe) are pinned by #2181's tests in `tests_numeric.zig` and `reader-exactness-gaps.scm`.

## Testing

- `zig build test`: 1622 pass, 0 fail (5 new regression tests: record field ladder 27/254/255/256, devfs stat, `seconds->time` bounds + timeout saturation, fiber channel timeout + unbounded-sleep park with fiber progress pinned so the test can't pass vacuously, 255-arg direct/tail/apply × exact/variadic + 256-param rejection).
- `bash tests/scheme/run-all.sh`: **674 pass, 0 fail, 0 timeout** (1 standard wasm-differential skip); R7RS suite 1395/1395. Three new suite files: `srfi170-devfs-1976.scm`, `srfi18-timeout-saturation-1983.scm`, `call-255-args-2185.scm`.
- All original issue repros re-run against the fixed binary, including the discriminating controls (26/27 field cliff, `/dev/fd`, 1.8e10/1.9e10 cliffs, 9.2e18/9.3e18, 253/254/255-param bisect).
- Cross-compiles clean for aarch64-windows, x86_64-linux, aarch64-openbsd (the other signed-`dev_t` platform), aarch64-netbsd, aarch64-freebsd, and wasm.

Closes #1973. Closes #1976. Closes #1983. Closes #2185.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by baijum — COMMENTED** [review]
## Bugbot review

Reviewed the full branch diff for PR #2186 (record sizing, signed `dev_t`, SRFI-18 timeout/time conversion, 255-arg call / 256-param lambda overflows).

**Result: no bugs found.**

The boundary hardenings look correct and well-pinned:
- `usize` widening for record alloc and call/rest-slot math
- bit-preserving `devToU64` for signed `dev_t`
- catchable rejection for `seconds->time` vs saturating timeouts
- clean `KP2001` for 256 fixed parameters

Regression coverage (unit + Scheme suites + re-enabled audit assertions) matches the claimed cliffs.

File: src/primitives_srfi18.zig [hunk 1]
```diff
@@ -601,8 +613,13 @@ fn threadSleepFn(args: []const Value) PrimitiveError!Value {
...
         if (seconds <= 0) return types.VOID;
-        const total_ns: u64 = @intFromFloat(@max(0.0, seconds * 1e9));
-        const deadline = fiber_mod.clockNs() + total_ns;
+        // lossyCast + saturating add (#1983): a duration too large for the
+        // u64 nanosecond clock (+inf.0, 1e300, an overflowed computation)
+        // saturates to a deadline that never fires -- an unbounded sleep,
+        // per the SRFI-18 "+inf.0 never times out" convention -- where the
+        // old @intFromFloat aborted the process uncatchably.
+        const total_ns: u64 = std.math.lossyCast(u64, @max(0.0, seconds * 1e9));
+        const deadline = fiber_mod.clockNs() +| total_ns;
```
[the remark this answers is not in the document]
Comment (baijum) [reply src/primitives_srfi18.zig]: Done in 462f73b2 — accepted, and agreed on the strongest form of the rationale: #1983 existed precisely because this conversion lived as three diverged copies, so a single edit point is the real hardening, not just tidiness. Two deliberate deviations from the proposed diff: the helper is conversion-only (each site keeps its own `clockNs() +|`, since the add is trivial and threadSleepFn's early-return structure stays readable), and its doc comment records why the two sites' historic `@max` spellings — before vs. after the multiply — were equivalent, so the unification is visibly behavior-preserving.

File: src/tests_filesystem.zig [hunk 2]
```diff
@@ -379,3 +379,30 @@ test "set-file-mode changes permissions" {
...
+        \\  (and (file-info? fi)
+        \\       (exact-integer? (file-info:device fi))
+        \\       (exact-integer? (file-info:rdev fi))
+        \\       (>= (file-info:device fi) 0)))
+    ));
+    // /dev/fd may be a broken /proc symlink on minimal containers, so only
+    // assert the stat cannot take the process down.
+    _ = try vm.eval(
+        \\(guard (e (#t 'stat-failed)) (file-info? (file-info "/dev/fd" #t)))
+    );
```
[the remark this answers is not in the document]
Comment (baijum) [reply src/tests_filesystem.zig:407]: Declining deliberately. `th.TestContext` is literally `GC.init` + `makeTestVM` bundled (testing_helpers.zig:54–70) — same objects, same lifetimes, no extra leak or error coverage — so there is no functional delta to gain. And the inline pattern is this file's uniform idiom: all 32 existing tests here use `makeTestVM` directly and none use TestContext, so converting only the new test would make it the file's one stylistic outlier. When the guideline's letter and the file's own convention conflict with zero behavioral stakes, matching the surrounding code wins; harmonizing whole test files onto TestContext would be a separate mechanical cleanup, not a rider on this fix.

File: src/tests_records.zig [hunk 3]
```diff
@@ -300,3 +300,90 @@ test "record-set! field survives full GC" {
...
+// (num_fields is a u8), so instantiating any record with >= 27 fields
+// (40 + 8*27 = 256) aborted the process with an uncatchable integer-overflow
+// panic. The type itself was created without complaint; only the first
+// instance died. A 256-field spec was worse: parseRecordSpec admitted it and
+// handleDefineRecordType's @intCast into the u8 panicked at DEFINITION time.
+test "wide records: 27 and 255 fields instantiate; 256 fields error cleanly" {
+    var gc = memory.GC.init(std.testing.allocator);
+    defer gc.deinit();
+    var vm = try th.makeTestVM(&gc);
+    defer vm.deinit();
```
[the remark this answers is not in the document]
Comment (baijum) [reply src/tests_records.zig:315]: Declining for the same reason as the tests_filesystem.zig comment: TestContext wraps the identical `GC.init` + `makeTestVM` sequence, and this file's dominant idiom (9 of its 12 tests) is the inline form the new test copies. No functional difference; file-local consistency kept.

File: src/vm_dispatch.zig [hunk 4]
```diff
@@ -459,7 +459,9 @@ pub fn runUntil(self: *VM, target_frame_count: usize, target_wind_count: usize)
                         self.registers[abs_base + 1 + rest_start] = try buildRestList(self.gc, self.registers[abs_base + 1 + rest_start .. abs_base + 1 + nargs]);
                     }
 
-                    const arg_count = if (func.is_variadic) func.arity + 1 else nargs;
+                    // arity is a u8 and 255 is legal, so the variadic +1 for
+                    // the rest-list slot must widen first (#2185).
+                    const arg_count: usize = if (func.is_variadic) @as(usize, func.arity) + 1 else nargs;
```
[the remark this answers is not in the document]
Comment (baijum) [reply src/vm_dispatch.zig:464]: Declining in this PR: vm_dispatch.zig was already 1,526 lines before this branch — the PR adds five, three of them comments — so the budget violation predates it and isn't something these bugfixes introduced. More importantly, this file is the VM's hottest path (the `runUntil` dispatch loop), and the policy itself says to split along natural seams, not to shed lines opportunistically; that deserves its own change with before/after `zig build bench` numbers, not a rider on a correctness fix. I've spun the split off as a tracked follow-up task so it isn't lost.

File: tests/scheme/audit/primitives_filesystem-audit.scm [hunk 5]
```diff
@@ -589,19 +589,19 @@
...
-;; (test-equal "(file-info-type (file-info '/dev/fd' #t))" 'directory (file-info-type (file-info "/dev/fd" #t)))
+;; #1976 regression: file-info used to abort (SIGABRT, uncatchable) on any
+;; path whose st_dev is negative when read as i32 — on macOS that is every
+;; entry under /dev, whose devfs st_dev is 2998493978 = -1296473318 as i32,
+;; via the old `.dev = @intCast(stat_buf.dev)`. Fixed by reinterpreting the
+;; bits as unsigned before widening (devToU64). /dev/fd has st_rdev = 0 and
+;; still aborted, which is what discriminated dev from rdev.
+(test-equal "(file-info-type (file-info '/dev/null' #t))" 'char-special (file-info-type (file-info "/dev/null" #t)))
+(test-equal "(file-info-device? (file-info '/dev/null' #t))" #t (file-info-device? (file-info "/dev/null" #t)))
+(test-equal "(file-info-type (file-info '/dev/fd' #t))" 'directory (file-info-type (file-info "/dev/fd" #t)))
```
[the remark this answers is not in the document]
Comment (baijum) [reply tests/scheme/audit/primitives_filesystem-audit.scm]: Fixed in 462f73b2, with one refinement over the suggested returns-or-raises guard: that shape would drop the assertion's whole reason for existing — /dev/fd is the audit's dev-vs-rdev discriminator (a directory with st_rdev = 0 whose stat still aborted pre-fix). The committed form accepts a catchable stat failure (broken /proc symlink, minimal environments — the hazard you're pointing at, which my other two new tests already guarded the same way) but still pins `'directory` wherever /dev/fd does resolve, which is every current CI leg. So the environment tolerance is gained without giving up the discrimination.
